lib/Sema/TypeCheckSwitchStmt.cpp

```diff
@@ -1137,7 +1137,104 @@ namespace {
         return false;
       }
     }
-    
+
+    /// \brief Estimate how big is the search space that exhaustivity
+    /// checker needs to cover, based on the total space and information
+    /// from the `switch` statement itself. Some of the easy situations
+    /// like `case .foo(let bar)` don't really contribute to the complexity
+    /// of the search so their sub-space sizes could be excluded from
+    /// consideration.
+    ///
+    /// \param total The total space to check.
+    /// \param covered The space covered by the `case` statements in the switch.
+    ///
+    /// \returns The size of the search space exhastivity checker has to check.
+    size_t estimateSearchSpaceSize(const Space &total, const Space &covered) {
+      switch (PairSwitch(total.getKind(), covered.getKind())) {
+        PAIRCASE(SpaceKind::Type, SpaceKind::Type) : {
+          return total.getType()->isEqual(covered.getType())
+                     ? 0
+                     : total.getSize(TC, DC);
+        }
+        PAIRCASE(SpaceKind::Type, SpaceKind::Disjunct)
+            : PAIRCASE(SpaceKind::Type, SpaceKind::Constructor) : {
+          if (!Space::canDecompose(total.getType(), DC))
+            break;
+
+          auto decomposition = Space::decompose(TC, DC, total.getType());
+          return estimateSearchSpaceSize(decomposition, covered);
+        }
+
+        PAIRCASE(SpaceKind::Disjunct, SpaceKind::Disjunct)
+            : PAIRCASE(SpaceKind::Disjunct, SpaceKind::Constructor) : {
+          auto &spaces = total.getSpaces();
+          return std::accumulate(spaces.begin(), spaces.end(), 0,
+                                 [&](size_t totalSize, const Space &space) {
+                                   return totalSize + estimateSearchSpaceSize(
+                                                          space, covered);
+                                 });
+        }
+
+        // Search space size computation is not commutative, because it
+        // tries to check if space on right-hand side is covering any
+        // portion of the "total" space on the left.
+        PAIRCASE(SpaceKind::Constructor, SpaceKind::Disjunct) : {
+          for (const auto &space : covered.getSpaces()) {
+            // enum E { case foo }
+            // func bar(_ lhs: E, _ rhs: E) {
+            //   switch (lhs, rhs) {
+            //     case (_, _): break
+            // }
+            if (total == space)
+              return 0;
+
+            if (!space.isSubspace(total, TC, DC))
+              continue;
+
+            if (estimateSearchSpaceSize(total, space) == 0)
+              return 0;
+          }
+          break;
+        }
+
+        PAIRCASE(SpaceKind::Constructor, SpaceKind::Constructor) : {
+          if (total.getHead().compare(covered.getHead()) != 0)
```

Comment:
Identifier has `==`, so you can just use that.

Reply:
Sounds good! I'll change here and in `minus` then, because it doesn't the same thing.